### PR TITLE
[extractor/youtube] Retry manifest refresh to avoid ending download early when live stream is temporarily offline

### PR DESCRIPTION
**Pull request by mzhou**
**IMPORTANT**: PRs without the template will be CLOSED

### Description of your *pull request* and other information

</details>

Currently when downloading a YouTube live stream, if it goes offline (e.g. due to a problem on the uploader's end), yt-dlp prints a warning like:

```
WARNING: [youtube] Cannot find refreshed manifest for format 140; please report this issue on  https://github.com/yt-dlp/yt-dlp/issues?q= , filling out the appropriate issue template. Confirm you are on the latest version using  yt-dlp -U
WARNING: [youtube] Cannot find refreshed manifest for format 299; please report this issue on  https://github.com/yt-dlp/yt-dlp/issues?q= , filling out the appropriate issue template. Confirm you are on the latest version using  yt-dlp -U
```

then considers the live stream over, proceeding to muxing.

This change makes the process of finding a refreshed manifest subject to `--extractor-retries` and `--retry-sleep extractor:`, so that downloading can continue without intervention as long as the live stream is back online within `extractor-retries * retry-sleep` seconds.

I tested this by starting a scheduled live stream, starting yt-dlp, then disconnecting the streaming software and confirming the retry messages from yt-dlp, and finally reconnecting the streaming software and observing that yt-dlp resumes downloading. I've also confirmed that the true end of the live stream is detected and yt-dlp proceeds to muxing fine.

Fixes #

<details open><summary>Template</summary> 

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8) and [ran relevant tests](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [x] Fix or improvement to an extractor (Make sure to add/update tests)
- [ ] New extractor ([Piracy websites will not be accepted](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#is-the-website-primarily-used-for-piracy))
- [ ] Core bug fix/improvement
- [ ] New feature (It is strongly [recommended to open an issue first](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#adding-new-feature-or-making-overarching-changes))

### CODE REVIEW

File: yt_dlp/extractor/youtube.py [hunk 1]
```diff
@@ -2617,22 +2617,32 @@ def refetch_manifest(format_id, delay):
...
-
-            f = next((f for f in formats if f['format_id'] == format_id), None)
-            if not f:
-                if not is_live:
-                    self.to_screen(f'{video_id}: Video is no longer live')
-                else:
-                    self.report_warning(
-                        f'Cannot find refreshed manifest for format {format_id}{bug_reports_message()}')
+            try:
+                for retry in self.RetryManager():
```
Comment (pukkandan) [inline yt_dlp/extractor/youtube.py:2628]:
```suggestion
                for retry in self.RetryManager(fatal=False):
```
Then remove the outer try

File: yt_dlp/extractor/youtube.py [hunk 2]
```diff
@@ -2617,22 +2617,32 @@ def refetch_manifest(format_id, delay):
...
+                        if not f:
+                            if not is_live:
+                                err_str = f'{video_id}: Video is no longer live'
+                            else:
+                                err_str = f'Cannot find refreshed manifest for format {format_id}{bug_reports_message()}'
+                            raise MpdFeedException(err_str)
+                        return f['manifest_url'], f['manifest_stream_number'], is_live
+                    except MpdFeedException as err:
+                        retry.error = err
+                        continue
```
Comment (pukkandan) [inline yt_dlp/extractor/youtube.py:2643]: No need to throw and catch error
```suggestion
                        if f:
                            return f['manifest_url'], f['manifest_stream_number'], is_live
                        elif is_live:
                            retry.error = f'Cannot find refreshed manifest for format {format_id}{bug_reports_message()}'
                        else:
                            retry.error = f'{video_id}: Video is no longer live'
```

**Review by pukkandan — CHANGES_REQUESTED** [review]

**Comment by pukkandan** [comment]
cc @Lesmiscore

**Review by Lesmiscore — APPROVED** [review]

File: yt_dlp/extractor/youtube.py [hunk 3]
```diff
@@ -2617,22 +2617,28 @@ def refetch_manifest(format_id, delay):
...
+
+                        f = next((f for f in formats if f['format_id'] == format_id), None)
+                        if f:
+                            return f['manifest_url'], f['manifest_stream_number'], is_live
+                        elif is_live:
+                            retry.error = f'Cannot find refreshed manifest for format {format_id}{bug_reports_message()}'
+                        else:
+                            retry.error = f'{video_id}: Video is no longer live'
+            except MpdFeedException:
                 return None
```
Comment (pukkandan) [inline yt_dlp/extractor/youtube.py:2641]: unused try except

File: yt_dlp/extractor/youtube.py [hunk 4]
```diff
@@ -2617,22 +2617,28 @@ def refetch_manifest(format_id, delay):
...
-            f = next((f for f in formats if f['format_id'] == format_id), None)
-            if not f:
-                if not is_live:
-                    self.to_screen(f'{video_id}: Video is no longer live')
-                else:
-                    self.report_warning(
-                        f'Cannot find refreshed manifest for format {format_id}{bug_reports_message()}')
+            try:
+                for retry in self.RetryManager(fatal=False):
+                    try:
```
Comment (pukkandan) [inline yt_dlp/extractor/youtube.py:2629]: ?

**Pull request merged**